### PR TITLE
chore(tracing): send traces to datadog

### PR DESCRIPTION
**Pull request by emjin**
Use a permanent collector instead of a local one for traces.

Test plan: see instructions in "How to add tracing" notion doc

### CODE REVIEW

**Comment by emjin** [comment]
PR comments for benchmarks have gotten much less sensitive. Here's the diff:

```
[semgrep.bench.zulip.std] 1.044x Baseline: 2.914, Latest: 3.041
[semgrep.bench.dropbox.std] 1.194x Baseline: 0.941, Latest: 1.124
[semgrep.bench.coinbase.std] 1.089x Baseline: 3.159, Latest: 3.441
[semgrep.bench.njs-django.std] 1.095x Baseline: 3.353, Latest: 3.671
[semgrep.bench.apache.std] 1.023x Baseline: 2.528, Latest: 2.585
[semgrep.bench.0c34.std] 1.181x Baseline: 1.005, Latest: 1.187
[semgrep.bench.lodash.std] 1.162x Baseline: 1.325, Latest: 1.540
[semgrep.bench.DVWA.std] 1.113x Baseline: 1.909, Latest: 2.125
[semgrep.bench.juice-shop.std] 1.053x Baseline: 5.817, Latest: 6.124
[semgrep.bench.Vulnerable-Flask-App.std] 1.076x Baseline: 2.467, Latest: 2.655
[semgrep.bench.pallets.std] 1.053x Baseline: 1.749, Latest: 1.842
[semgrep.bench.coolMenu.std] 1.173x Baseline: 0.953, Latest: 1.117
[semgrep.bench.t00sh.std] 1.094x Baseline: 3.456, Latest: 3.781
[semgrep.bench.grpc.std] 1.118x Baseline: 1.309, Latest: 1.464
[semgrep.bench.kotlinPrefilterTest.std] 1.047x Baseline: 2.981, Latest: 3.122
Average: 1.101x, Min: 1.023x, Max: 1.194x
Total Baseline: 35.868 s, Latest: 38.820 s
```

But develop has slowed down as well, so I think we're not seeing a significant perf drop! Need to still investigate why datadog doesn't seem to have received the per-file traces.

```
[semgrep.bench.zulip.std] 1.036x Baseline: 2.901, Latest: 3.006
[semgrep.bench.dropbox.std] 1.170x Baseline: 0.943, Latest: 1.103
[semgrep.bench.coinbase.std] 1.084x Baseline: 3.166, Latest: 3.432
[semgrep.bench.njs-django.std] 1.062x Baseline: 3.337, Latest: 3.542
[semgrep.bench.apache.std] 0.997x Baseline: 2.527, Latest: 2.520
[semgrep.bench.0c34.std] 1.169x Baseline: 1.015, Latest: 1.186
[semgrep.bench.lodash.std] 1.150x Baseline: 1.342, Latest: 1.544
[semgrep.bench.DVWA.std] 1.112x Baseline: 1.902, Latest: 2.114
[semgrep.bench.juice-shop.std] 1.031x Baseline: 5.788, Latest: 5.970
[semgrep.bench.Vulnerable-Flask-App.std] 1.058x Baseline: 2.497, Latest: 2.640
[semgrep.bench.pallets.std] 1.057x Baseline: 1.752, Latest: 1.852
[semgrep.bench.coolMenu.std] 1.158x Baseline: 0.955, Latest: 1.106
[semgrep.bench.t00sh.std] 1.101x Baseline: 3.452, Latest: 3.800
[semgrep.bench.grpc.std] 1.099x Baseline: 1.313, Latest: 1.443
[semgrep.bench.kotlinPrefilterTest.std] 1.010x Baseline: 3.006, Latest: 3.036
Average: 1.086x, Min: 0.997x, Max: 1.170x
Total Baseline: 35.894 s, Latest: 38.294 s
Reading ci_small_repos_findings.json
Reading /home/runner/work/semgrep/semgrep/perf/snapshots/ci_small_repos_baseline.json
```

**Comment by emjin** [comment]
Semgrep Pro check is unrelated

File: libs/tracing/unix/Tracing.ml
Comment (cgdolan) [inline libs/tracing/unix/Tracing.ml]: What do you think about having a flag for setting the endpoint too? I'm imagining this would default to prod (`https://telemetry.semgrep.dev`) if not specified, otherwise whatever the user passes as the value.

This would enable three use cases pretty cleanly:
1. Customers want to send us the trace, it goes to prod: `semgrep <scan|ci> --trace`
2. We want to send a trace for benchmarks, it goes to dev: `semgrep <scan|ci> --trace --trace-endpoint https://telemetry.dev2.semgrep.dev`
3. We want to send trace for developing locally, it goes to locally running trace collector (Jaeger): `semgrep <scan|ci> --trace --trace-endpoint localhost:4138`

Comment (emjin) [reply libs/tracing/unix/Tracing.ml]: I have an env var; would you prefer a flag?

Comment (emjin) [reply libs/tracing/unix/Tracing.ml]: (Flags are just a bit annoying to add because you have to thread them through semgrep cli and then osemgrep)

File: libs/tracing/unix/Tracing.ml
Comment (cgdolan) [inline libs/tracing/unix/Tracing.ml:58]:
```suggestion
let default_endpoint = "https://telemetry.semgrep.dev"
```

File: scripts/run-benchmarks.sh
Comment (cgdolan) [inline scripts/run-benchmarks.sh:34]:
```suggestion
SEMGREP_OTEL_ENDPOINT=https://telemetry.dev2.semgrep.dev pipenv run python3 ../perf/run-benchmarks --config $config_path --std-only --trace --save-to timing2.json --save-findings-to ci_small_repos_findings.json
```

File: libs/tracing/unix/Tracing.ml
Comment (cgdolan) [reply libs/tracing/unix/Tracing.ml:58]: Suggest making prod the default and requiring explicit, additional configuration to send to non-prod.

File: libs/tracing/unix/Tracing.ml
Comment (cgdolan) [inline libs/tracing/unix/Tracing.ml:59]: I wonder if this is necessary given the client supports configuration via the `OTEL_EXPORTER_OTLP_ENDPOINT` env var? It might be that instantiating the client with a default value and then overriding with this env var just works (tm) and that we don't actually need another env var here. What do you think?

File: libs/tracing/unix/Tracing.ml
Comment (cgdolan) [inline libs/tracing/unix/Tracing.ml:51]: Looks like these docs need updating (TODO for endpoint) and possibly instructions on sending to another endpoint with `OTEL_EXPORTER_OTLP_ENDPOINT` if this no longer works after setting the default value in the client config.

**Review by cgdolan — APPROVED** [review]
Overall, looks great! Approving since my comments are all nits and I trust you'll do something reasonable. Feel free to resolve and merge!

File: libs/tracing/unix/Tracing.ml
Comment (kopecs) [reply libs/tracing/unix/Tracing.ml:59]: We do need another. See https://github.com/imandra-ai/ocaml-opentelemetry/blob/9e2fc9445b321546fe3e6d874bbd907da25c1a6b/src/client-ocurl/config.ml#L35 (default here: https://github.com/imandra-ai/ocaml-opentelemetry/blob/9e2fc9445b321546fe3e6d874bbd907da25c1a6b/src/client-ocurl/common_.ml#L16-L21)

File: libs/tracing/unix/Tracing.ml
Comment (kopecs) [inline libs/tracing/unix/Tracing.ml]:
```suggestion
  let config = Opentelemetry_client_ocurl.Config.make ~url () in
  Opentelemetry_client_ocurl.with_setup ~config () @@ fun () ->
```

**Review by kopecs — COMMENTED** [review]
Seems like this needs the comment updates Chris commented on (top of Tracing.ml) but otherwise pretty good.

Do the per-target spans have the correct parent right now? If so, lgtm.

File: libs/tracing/unix/Tracing.ml
Comment (cgdolan) [reply libs/tracing/unix/Tracing.ml:59]: Ah, bummer. Thanks for confirming.

**Comment by emjin** [comment]
The per-target spans work for semgrep but not for semgrep-pro, which I've set up similarly. I'm going to look into that before merging.

**Comment by kopecs** [comment]
Can we just use with_span then :)

File: libs/tracing/unix/Tracing.ml
Comment (cgdolan) [reply libs/tracing/unix/Tracing.ml]: An env var is totally fine, my feedback was really that we should have some mechanism for customizing the endpoint.

File: libs/tracing/unix/Tracing.ml
Comment (emjin) [reply libs/tracing/unix/Tracing.ml:58]: I'm going to leave it as dev for now since it's still in progress but I'll make a ticket to switch it

Comment (emjin) [reply libs/tracing/unix/Tracing.ml:58]: https://linear.app/semgrep/issue/SAF-869/switch-default-endpoint-to-prod

File: libs/tracing/unix/Tracing.ml
Comment (emjin) [reply libs/tracing/unix/Tracing.ml:51]: Good catch, thanks!

**Pull request merged**
